Source/engine/random.hpp

```diff
@@ -136,6 +136,83 @@ class DiabloGenerator {
 	}
 };
 
+/** Adapted from https://prng.di.unimi.it/xoshiro128plusplus.c */
+class xoshiro128plusplus {
+private:
+	static uint32_t rotl(const uint32_t x, int k)
+	{
+		return (x << k) | (x >> (32 - k));
+	}
+
+	uint32_t s[4];
+
+public:
+	void seed(uint64_t value)
+	{
+		uint64_t rand1 = next_split(value);
+		uint64_t rand2 = next_split(rand1);
+		s[0] = static_cast<uint32_t>(rand1);
+		s[1] = static_cast<uint32_t>(rand2);
+		s[2] = static_cast<uint32_t>(rand1 >> 32);
+		s[3] = static_cast<uint32_t>(rand2 >> 32);
+	}
+
+	void seed(uint32_t value)
+	{
+		s[0] = next_split(value);
+		s[1] = next_split(s[0]);
+		s[2] = next_split(s[1]);
+		s[3] = next_split(s[2]);
+	}
+
+	uint32_t next()
+	{
+		const uint32_t result = rotl(s[0] + s[3], 7) + s[0];
+
+		const uint32_t t = s[1] << 9;
+
+		s[2] ^= s[0];
+		s[3] ^= s[1];
+		s[1] ^= s[2];
+		s[0] ^= s[3];
+
+		s[2] ^= t;
+
+		s[3] = rotl(s[3], 11);
+
+		return result;
+	}
+
+private:
+	/** Adapted from https://prng.di.unimi.it/splitmix64.c */
+	static uint64_t next_split(uint64_t x)
+	{
+		uint64_t z = (x += 0x9e3779b97f4a7c15);
+		z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9;
+		z = (z ^ (z >> 27)) * 0x94d049bb133111eb;
+		return z ^ (z >> 31);
+	}
+
+	/** Adapted from https://prng.di.unimi.it/splitmix64.c using fmix32 from MurmurHash3 */
+	static uint32_t next_split(uint32_t x)
+	{
+		uint32_t z = (x += 0x9e3779b9);
+		z = (z ^ (z >> 16)) * 0x85ebca6b;
+		z = (z ^ (z >> 13)) * 0xc2b2ae35;
+		return z ^ (z >> 16);
```

Comment:
Is this from <https://github.com/umireon/my-random-stuff/blob/e7b17f992955f4dbb02d4016682113b48b2f6ec1/xorshift/splitmix32.c>? Even though it's public domain/CC0 probably safest to link back/reference the paper those constants were derived from. I also see other sources (<https://github.com/bryc/code/blob/master/jshash/PRNGs.md>) with different shift values and constants that offer some statistical benefit but doubt it makes much difference here.

Reply:
Actually, no. My only sources were the SplitMix64 implementation and MurmurHash3.
https://github.com/aappleby/smhasher/blob/61a0530f28277f2e850bfc39600ce61d02b518de/src/MurmurHash3.cpp#L68

That `fmix32()` function doesn't include the first constant, but I guess that was derived from the golden ratio. IIRC, I settled on it after Googling for the meaning behind the corresponding constant used by the SplitMix64 algorithm.

Reply:
I suppose, if you look at the source code for SplitMix64, you'll see it was written in 2015 and provides the following link to a research article from 2014. It's the same article with the same authors that's referenced by Kaito Udagawa's source file that you discovered. http://dx.doi.org/10.1145/2714064.2660195

Given that, I guess it is okay to go ahead an reach past SplitMix64 to reference the article directly, although I'm iffy about giving attribution to Kaito Udagawa since his implementation came later, and I didn't actually use it.

Reply:
All good, the kaito implementation was the only one I found with that specific set of constants/shifts operating on a 32 bit state but I didn't look particularly hard 😅 .